### DISCUSSION TITLE
LaTeX Printing fold_excess_bracing Option

### DISCUSSION THREAD

**Discussion by Craie** [sympy/sympy#24779] [Ideas]
I'd like to suggest a `fold_excess_bracing` option for LaTeX printing.  I'm using this system with PrairieLearn to test math students and having them add lots of {} around single symbols is not only hard but unnecessary.

I.E. When I do `latex(Sum(a[i], (i, k, n)))` I get `'\\sum_{i=k}^{n} {a}_{i}'`  but the curly braces around `n`, `a`, and `i` are all unnecessary.  Only the ones around the `i=k` are needed.  Could there be an option to whittle down the extra bracing?

(I teach this issue in expression tree printing in my basic data structures course, so I know it can be done.  The question is will this suggestion be accepted...)

**Comment by oscarbenjamin** [comment]
I don't see why it wouldn't be accepted. I'm not entirely sure how this works in sympy's printing code but there are tests for when things need brackets in different contexts. I'm not sure if what they do is equivalent to what you are suggesting. I suppose that with LaTeX the application of curly brackets is not as well developed because LaTeX is usually used (in the SymPy context) as an intermediate format.

**Reply by Craie** [reply]
I've implemented a kludge that uses a regular expression replacement to remove curly braces with single characters inside.  But it is less than ideal.

I'm really hoping to find an equivalence matcher that would let me say that two formulas were equal -- regardless of commutativity and such.  But checking the `latex` strings is about as close as I've found so far.

**Reply by oscarbenjamin** [reply]
I'm not sure I understand what you mean by formulas being equal and why checking the latex would help.

Improving the latex output to remove unnecessary curly brackets is good in any case though although we should do it conservatively because it is more important that the latex compiles than that it looks nice in itself.

**Reply by Craie** [reply]
True to the last point.  That's why I was only suggesting it as an option like folding fractions is an option.

By equivalence I meant like two sums that run between the same two bounds but with a different index of summation.  Or like `a+b` would be the same as if someone wrote `b+a`.  Right now, my check is an exact string match and so I can't account for the commutativity or variable naming.

**Reply by oscarbenjamin** [reply]
SymPy already canonicalises the order of terms in an Add:
```
In [1]: x + y
Out[1]: x + y

In [2]: y + x
Out[2]: x + y

In [3]: x + y == y + x
Out[3]: True
```
If you're looking for a way to export SymPy expressions to something that is easier to parse then I suggest:
```python
In [4]: srepr(x + y)
Out[4]: "Add(Symbol('x'), Symbol('y'))"
```

**Reply by Craie** [reply]
Interesting.  I hadn't seen those features yet.  Thank you for the information!  I'll look further into that for sure.

**Reply by sylee957** [reply]
> I'm really hoping to find an equivalence matcher that would let me say that two formulas were equal -- regardless of commutativity and such. But checking the latex strings is about as close as I've found so far.

I think that you would need to tune sympy's latex parser to do that because exact strings are too 'low level' to test out the equivalence.

**Reply by Craie** [reply]
I've written a pair of regex to help in canonicalizing the problem I'm working on.  The main trouble was the summation index which can be arbitrarily anything.  I just preparsed their input before the grader got to it -- a step PrairieLearn allows, thank goodness!

**Comment by sylee957** [comment]
The only thing complicates about implementing that is we need to understand the mechanism how braces can be hidden or not.

For example, `x^123` prints out $x^123$ so it looks like we need to restrict it to single character or 'token' at least. 
But I'm not sure which extent it can be always done or not, because I haven't read full technical treatment of LaTeX

And this is one of the most common mistakes (even among the sympy developers) to typeset the LaTeX in docs.

**Reply by Craie** [reply]
The rule goes all the way back to the underlying TeX and is, indeed, that exactly one character is assumed unless bracing is used.  So you can just check the length of what is in the node before printing and, if one, don't output {}.

**Reply by sylee957** [reply]
I don't think that we need to add additional options for printing this though. We can always hide the brackets if unnecessary.
